util/src/main/java/org/datacommons/util/StringUtil.java

```diff
@@ -124,25 +133,27 @@ public static String getValidISO8601DatePattern(String dateValue) {
 
   public static LocalDateTime getValidISO8601Date(String dateValue) {
     // TODO: handle the extra date patterns
-    for (String pattern : DATE_PATTERNS) {
-      try {
-        DateTimeFormatter dateFormat =
-            new DateTimeFormatterBuilder()
-                .appendPattern(pattern)
-                .parseDefaulting(ChronoField.DAY_OF_MONTH, 1)
-                .parseDefaulting(ChronoField.MONTH_OF_YEAR, 1)
-                .parseDefaulting(ChronoField.HOUR_OF_DAY, 0)
-                .toFormatter(Locale.ENGLISH);
-        return LocalDateTime.parse(dateValue, dateFormat);
-      } catch (DateTimeParseException ex) {
-        // Pass through
+    for (Entry<String, Set<Integer>> entry : DATE_PATTERNS.entrySet()) {
+      if (entry.getValue().contains(dateValue.length())) {
```

Comment:
From runtime perspective, I slightly prefer your prior approach of looking up a single map instead of looping over all patterns, and then doing set lookups.

I was originally mistaken that the lengths were just strlen, and thus my suggestion to "compute" it.

So, we could compute len->list-of-patterns from the map you have.  Or with lesser code, just make the Map as a List of pairs so we try YYYY first, the YYYYMM (and that's the most common).

Reply:
I was going for the lesser code with my revision, so I will implement the Map as a List of pairs so that we prioritize common cases and maintain the performance gain with less speed. (also would make the tests deterministic again and allow for no diffs to the tests)

Computing len>list-of-patterns map on the first call to the function and then using that map for every call, while marginally faster, involves more code as you mentioned. The optimization focus of this change is to have as few exceptions thrown as possible, so I will go with the less code option above. We can revise as needed.

Reply:
Moved to List of pairs.